File: classic-puzzle-easy/go/logic-gates.go
package main

import (
	"fmt"
	"strings"
)

func main() {
	var n int
	fmt.Scan(&n)

	var m int
	fmt.Scan(&m)

	inputSignals := make(map[string]string) // Dictionary
	for i := 0; i < n; i++ {
		var inputName, inputSignal string
		fmt.Scan(&inputName, &inputSignal)
		inputSignals[inputName] = inputSignal
	}

	gateOperations := map[string]func(bool, bool) bool{
		"AND":  func(b1, b2 bool) bool { return b1 && b2 },
		"OR":   func(b1, b2 bool) bool { return b1 || b2 },
		"XOR":  func(b1, b2 bool) bool { return b1 != b2 },
		"NAND": func(b1, b2 bool) bool { return !(b1 && b2) },
		"NOR":  func(b1, b2 bool) bool { return !(b1 || b2) },
		"NXOR": func(b1, b2 bool) bool { return b1 == b2 },
	}

	outputs := make([]string, m)
	for i := 0; i < m; i++ {
		var outputName, gateType, inputName1, inputName2 string
		fmt.Scan(&outputName, &gateType, &inputName1, &inputName2)

		signal1 := inputSignals[inputName1]
		signal2 := inputSignals[inputName2]
		var outputSignal strings.Builder

		for j := range signal1 {
			b1 := signal1[j] == '-'
			b2 := signal2[j] == '-'

			if gateOperations[gateType](b1, b2) {
				outputSignal.WriteRune('-')
			} else {
				outputSignal.WriteRune('_')
			}
		}

		outputs[i] = fmt.Sprintf("%s %s", outputName, outputSignal.String())
	}

	for _, output := range outputs {
		fmt.Println(output)
	}
}

File: classic-puzzle-easy/python/logic-gates.py
import sys
import math

n = int(input())
m = int(input())

# Define gate operations
gate_operations = {
    'AND': lambda b1, b2: b1 and b2,
    'OR': lambda b1, b2: b1 or b2,
    'XOR': lambda b1, b2: b1 ^ b2,
    'NAND': lambda b1, b2: not (b1 and b2),
    'NOR': lambda b1, b2: not (b1 or b2),
    'NXOR': lambda b1, b2: not (b1 ^ b2)
}

# Read input signals
input_signals = {}
for i in range(n):
    input_name, input_signal = input().split()
    input_signals[input_name] = [x == '-' for x in input_signal]

# Read output specifications and compute results
outputs = []
for _ in range(m):
    output_name, gate_type, input_name_1, input_name_2 = input().split()
    signal1 = input_signals[input_name_1]
    signal2 = input_signals[input_name_2]
    output_signal = [gate_operations[gate_type](b1, b2) for b1, b2 in zip(signal1, signal2)]
    output_signal_str = ''.join('-' if bit else '_' for bit in output_signal)
    outputs.append(f"{output_name} {output_signal_str}")

# Print the outputs
for output in outputs:
    print(output)
